### PR TITLE
Var Type Name Validators

### LINKED ISSUE #1351: Creating duplicate messages
#1350 introduces a UX issue where users would see a discrepancy between the first time and additional times they try to create a new message `message1` using any of the broadcast blocks in the flyout.

E.g.
1. Open a new project
2. Use one of the broadcast blocks in the flyout to create a new message `message1`
3. Repeat step 2.

Performing step 2 looks like a no-op to the user, while performing step 3 results in the following pop-up:

image

The discrepancy here is confusing to the user. What's happening behind the scenes is that performing step 2 the first time transforms the potential variable (a concept introduced by #1350) into a real variable of the same name (this same thing would happen if any of the broadcast blocks are dragged from the flyout into the workspace). Repeating step 2 yields an error because now, a variable with the same name and type already exists.

----------------------------------------------------------------------------------------------------

### Proposed Solution:
Change validator for broadcast messages to not complain when a user tries to create a new message with the same name as an existing broadcast message. In Scratch 2.0, users can create broadcast messages with the same name as many times as they want. The result of performing this action appears to be a no-op to the user.

### PR DESCRIPTION
**Pull request by kchadha**
### Resolves

Resolves #1351 
And some residual todos from #1244

### Proposed Changes

The proposed changes are a major refactor of the code that handles creation and renaming of variables. The changes include:

- The validation of variable names according to type is more explicit now. There are two functions, one that handles name validation for broadcast messages and one that handles name validation for both scalar variables and lists.

- The user can now 'create' (results in a no-op) new broadcast messages that have the same name as one that already exists as a broadcast message. This issue is discussed in #1351, and this new behavior is consistent with Scratch 2.0.

- The prompt for creating or renaming variables no longer recurses when the supplied name already exists. This behavior was already not present in the scratch-gui due to a bug, but this new behavior is consistent with Scratch 2.0.

- The validators for scalar variables and lists no longer merge runs of whitespace in the middle of a variable name (e.g. 'my        variable' and 'my  variable' can co-exist as two separate scalar variable names). White space at the beginning or end of a scalar variable or list name is still trimmed. This new behavior is consistent with Scratch 2.0.

### Reason for Changes

The issues listed above, code cleanup, and consistency with Scratch 2.0 behavior.

### Test Coverage

Existing tests pass.

### CODE REVIEW

File: core/variables.js [hunk 1]
```diff
@@ -259,72 +259,182 @@ Blockly.Variables.realizePotentialVar = function(varName, varType, potentialVarW
...
+  } else {
+    console.warn('Encountered unexpected type "' + opt_type + '" in call to ' +
+        'Blockly.Variables.createVariable');
+    return;
   }
+  var validate = Blockly.Variables.nameValidator_.bind(null, opt_type);
   // This function needs to be named so it can be called recursively.
-  var promptAndCheckWithAlert = function(defaultName) {
-    Blockly.Variables.promptName(newMsg, defaultName,
+  //var promptAndCheckWithAlert = function(defaultName) {
```
Comment (rachel-fenichel) [inline core/variables.js]: Remove commented-out code.

File: core/variables.js [hunk 2]
```diff
@@ -259,72 +259,182 @@ Blockly.Variables.realizePotentialVar = function(varName, varType, potentialVarW
...
-      }, opt_type, modalTitle);
-  };
-  promptAndCheckWithAlert('');
+      }, modalTitle);
+};
+
+/**
+ * This function provides a common interface for variable name validation agnostic
+ * of type. This is so that functions like Blockly.Variables.createVariable and
+ * Blockly.Variables.renameVariable can call a single function (with a single type
```
Comment (rachel-fenichel) [inline core/variables.js]: "type" is duplicated: on the end of this line and the start of the next line.

File: core/variables.js [hunk 3]
```diff
@@ -259,72 +259,182 @@ Blockly.Variables.realizePotentialVar = function(varName, varType, potentialVarW
...
+ *     is determined to be invalid/in-use, and the calling function should not
+ *     proceed with creating or renaming the variable.
+ * @private
+ */
+Blockly.Variables.nameValidator_ = function(type, text, workspace, opt_callback) {
+  // The validators for the different variable types require slightly different arguments.
+  // For broadcast messages, if a broadcast message of the provided name already exists,
+  // the validator needs to call a function that updates the selected
+  // field option of the dropdown menu of the block that was used to create the new message.
+  // For scalar variables and lists, the validator needs a retry function that is responsible
```
Comment (rachel-fenichel) [inline core/variables.js]: I don't see a retry function.  Is this comment out of date?

File: core/variables.js [hunk 4]
```diff
@@ -337,77 +447,55 @@ Blockly.Variables.createVariable = function(workspace, opt_callback, opt_type) {
...
+  } else if (varType == Blockly.SCALAR_VARIABLE_TYPE) {
+    promptMsg = Blockly.Msg.RENAME_VARIABLE_TITLE;
+    modalTitle = Blockly.Msg.RENAME_VARIABLE_MODAL_TITLE;
+  } else {
+    // Only lists and variables can be renamed... this is an error
+    console.warn('Encountered unexpected variable type ' + varType +
+        'when attempting to rename a variable.');
+    return;
+  }
+  var validate = Blockly.Variables.nameValidator_(varType);
```
Comment (rachel-fenichel) [inline core/variables.js]: Line 286 is 
`  var validate = Blockly.Variables.nameValidator_.bind(null, opt_type);`

Why is this one handled differently?

**Review by rachel-fenichel — COMMENTED** [review]
Structurally this looks fine.

You're making it very hard to create variables of any types other than the 3 in scratch.  This is a problem as a library, though people will be able to easily see where they need to add code to resolve it.

File: core/variables.js (hunk 1 above)
Comment (kchadha) [reply core/variables.js]: done.

File: core/variables.js (hunk 2 above)
Comment (kchadha) [reply core/variables.js]: done.

File: core/variables.js (hunk 3 above)
Comment (kchadha) [reply core/variables.js]: fixed.

File: core/variables.js (hunk 4 above)
Comment (kchadha) [reply core/variables.js]: Oops! that was unintentional. Fixed.

**Comment by kchadha** [comment]
I agree that the proposed changes make it harder to create variables of any other type, and that this is not necessarily a good thing.

One of the things I was debating about was whether or not to keep the functionality of defaulting to the scalar variable type if type information isn't provided or isn't one of the three we already know about (instead of having a warning), but I think the warning will help make it clear to a developer that we expect some sort of name validation, and that it can be type specific or it can follow along with the scalar variable/list naming conventions.

I think in general, this refactor makes it clearer what the rules of each type of variable are, and gives examples of the different things you can do with the name validators. If there are other things I can do to make this more library/developer friendly, within the scope of these changes, definitely let me know.

**Comment by paulkaplan** [comment]
@rachel-fenichel @kchadha and I discussed this a little bit. @kchadha correct me if any of this doesn't seem right:

I think it makes sense to have a "default" validation behavior (probably the case-insensitive scalar one) to enable people to experiment with adding new variable types when using scratch-blocks as a library without having to make their own fork. I still don't have a great vision of how people might use custom variable types without having to fork to make other edits to support it, but it does seem like it if it is just an issue of validation behavior, having a default, particularly since you can't provide your own, is a good idea.

One caveat is we should make sure we are falling back to the _behavior_ of scalar variable types, never actually _using_ the scalar variable type when we don't know what else to use. Doing that would cause very hard to debug issues later down the line. That used to be an issue in the field_dropdown before all the recent work you all did, and it doesn't look like anything like that was introduced here, but just wanted to call it out.

Does that seem right to you @kchadha and @rachel-fenichel?

TL;DR I think it is ok to use the scalar name validation behavior for new variable types we don't know have. That means
- remove [this warning in create](https://github.com/kchadha/scratch-blocks/blob/7a63cb1237f7df655b8ecc2c6e431b6f9659c579/core/variables.js#L282-L283) and using the scalar variable messages instead
- remove [this warning in rename](https://github.com/kchadha/scratch-blocks/blob/7a63cb1237f7df655b8ecc2c6e431b6f9659c579/core/variables.js#L462-L463) and using the scalar variable messages instead
- remove [this warning in the validator](https://github.com/kchadha/scratch-blocks/blob/7a63cb1237f7df655b8ecc2c6e431b6f9659c579/core/variables.js#L366-L367) and use the scalar variable validator
- remove [this early return in the scalar validator](https://github.com/kchadha/scratch-blocks/blob/7a63cb1237f7df655b8ecc2c6e431b6f9659c579/core/variables.js#L420-L424)

**Comment by rachel-fenichel** [comment]
@paulkaplan @kchadha I like that solution.

**Comment by kchadha** [comment]
@paulkaplan's suggestion sounds good to me.
One remaining question I have is: what should be the behavior if the type is not provided?

**Comment by rachel-fenichel** [comment]
I suggest defaulting to the empty string; once you switch scalar variables to 'scalar' you could add a warning if the type is empty: "did you mean to do that?"

**Comment by kchadha** [comment]
@rachel-fenichel @paulkaplan, I made the requested changes. The only thing I left in is a warning/early return if the developer has called Blockly.renameVariable on a broadcast message, since we explicitly disallow renaming broadcast messages. Let me know if anything needs to change.
Thanks for reviewing!

**Review by rachel-fenichel — APPROVED** [review]
LG

**Review by paulkaplan — APPROVED** [review]
LG
